modules/foreman_bookmark.py

```diff
@@ -0,0 +1,177 @@
+version_added: "2.4"
+author:
+  - "Bernhard Hopfenmueller (@Fobhep) ATIX AG"
+requirements:
+  - "nailgun >= 0.29.0"
+  - "ansible >= 2.3"
+options:
+  server_url:
+    description:
+      - URL of Foreman server
+    required: true
+  username:
+    description:
+      - Username on Foreman server
+    required: true
+  password:
+    description:
+      - Password for user accessing Foreman server
+    required: true
+  verify_ssl:
+    description:
+      - Verify SSL of the Foreman server
+    required: false
+    default: true
+    type: bool
+  name:
+    description:
+      - Name of the bookmark
+    required: true
+  controller:
+    description:
+      - Controller for the bookmark
+    required: true
+  public:
+    description:
+      - Make bookmark available for all users
+    required: false
+    default: true
+    type: bool
+  query:
+    description:
+      - Query of the bookmark
+  state:
+    description:
+      - State of the bookmark
+    default: present
+    choices:
+      - present
+      - present_with_defaults
+      - absent
+'''
+
+EXAMPLES = '''
+- name: "Create a Bookmark"
+  foreman_bookmark:
+    username: "admin"
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    name: "recent"
+    controller: "job_invocations"
+    query: "started_at > '24 hours ago'"
+    state: present_with_defaults
+
+- name: "Update a Bookmark"
+  foreman_bookmark:
+    username: "admin"
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    name: "recent"
+    controller: "job_invocations"
+    query: "started_at > '12 hours ago'"
+    state: present
+
+- name: "Delete a Bookmark"
+  foreman_bookmark:
+    username: "admin"
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    name: "recent"
+    controller: "job_invocations"
+    state: absent
+'''
+
+RETURN = ''' # '''
+
+try:
+    from nailgun.entities import (
+        Bookmark,
+    )
+
+    from ansible.module_utils.ansible_nailgun_cement import (
+        find_bookmark,
+        find_locations,
+        find_organizations,
+        naildown_entity_state,
+        sanitize_entity_dict,
+    )
+except ImportError:
+    pass
+
+
+# This is the only true source for names (and conversions thereof)
+name_map = {
+    'name': 'name',
+    'controller': 'controller',
+    'public': 'public',
+    'query': 'query',
+}
+
+
+def main():
+    module = ForemanEntityAnsibleModule(
+        argument_spec=dict(
+            name=dict(required=True),
+            controller=dict(),
+            public=dict(defaut='true', type='bool'),
+            query=dict(),
+            state=dict(default='present', choices=[
+                       'present_with_defaults', 'present', 'absent']),
+        ),
+        required_if=(
+            ['state', 'present', ['name', 'query', 'controller']],
+            ['state', 'present_with_defaults', ['name', 'query', 'controller']],
+            ['state', 'absent', ['name', 'controller']],
```

Comment:
You can remove `name` as it required anyway. Maybe also controller, I'm not sure if that is actually required.

Reply:
You can create bookmarks with identical names. In order to be unique a bookmark needs for specification both controller and name

Reply:
fixed